README.md
# hordelib

[![PyPI version](https://badge.fury.io/py/hordelib.svg?branch=main&nocache=1)](https://badge.fury.io/py/hordelib)
[![Downloads](https://pepy.tech/badge/hordelib)](https://pepy.tech/project/hordelib)
![GitHub license](https://img.shields.io/github/license/jug-dev/hordelib)

[![Build](https://github.com/jug-dev/hordelib/actions/workflows/maintests.yml/badge.svg?branch=main)](http://hordelib.s3-website-eu-west-1.amazonaws.com/)
[![Test Images](https://badgen.net/badge/main/latest-images/blue?icon=awesome)](http://hordelib.s3-website-eu-west-1.amazonaws.com/)
[![Test Images](https://badgen.net/badge/develop/latest-images/blue?icon=awesome)](http://hordelib.s3-website-eu-west-1.amazonaws.com/unstable)

`hordelib` is a thin wrapper around [ComfyUI](https://github.com/comfyanonymous/ComfyUI) primarily to enable the [AI Horde](https://aihorde.net/) to run inference pipelines designed visually in the ComfyUI GUI.

The developers of this project can be found in the AI Horde Discord server: [https://discord.gg/3DxrhksKzn](https://discord.gg/3DxrhksKzn)

NOTE: This project is in early development and is not yet in use by Stable Horde.

## Purpose

The goal here is to be able to design inference pipelines in the excellent ComfyUI, and then call those inference pipelines programmatically. Whilst providing features that maintain compatibility with the existing horde implementation.

## Installation

If being installed from pypi, use a requirements file of the form:
```
--extra-index-url https://download.pytorch.org/whl/cu117
hordelib

...your other dependencies...
```

## Usage

Horde payloads can be processed simply with (for example):

```python
import os
import hordelib
hordelib.initialise()

from hordelib.horde import HordeLib
from hordelib.shared_model_manager import SharedModelManager

# Wherever your models are
os.environ["AIWORKER_CACHE_HOME"] = "f:/ai/models"

generate = HordeLib()
SharedModelManager.loadModelManagers(compvis=True)
SharedModelManager.manager.load("Deliberate")

data = {
    "sampler_name": "k_dpmpp_2m",
    "cfg_scale": 7.5,
    "denoising_strength": 1.0,
    "seed": 123456789,
    "height": 512,
    "width": 512,
    "karras": True,
    "tiling": False,
    "hires_fix": False,
    "clip_skip": 1,
    "control_type": None,
    "image_is_control": False,
    "return_control_map": False,
    "prompt": "an ancient llamia monster",
    "ddim_steps": 25,
    "n_iter": 1,
    "model": "Deliberate",
}
pil_image = generate.basic_inference(data)
pil_image.save("test.png")
```

Note that `hordelib.initialise()` will erase all command line arguments from argv. So make sure you parse them before you call that.

See `tests/run_*.py` for more standalone examples.

### Logging

If you don't want `hordelib` to setup and control the logging configuration initialise with:

```python
import hordelib
hordelib.initialise(setup_logging=False)
```

## Development

Requirements:
- `git` (install git)
- `tox` (`pip install tox`)
- Set the environmental variable `AIWORKER_CACHE_HOME` to point to a model directory.

Note the model directory must currently be in the original AI Horde directory structure:
```
<AIWORKER_CACHE_HOME>\
   nataili\
        clip\
        codeformer\
        compvis\
            Deliberate.ckpt
            ...etc...
        controlnet\
        embeds\
        esrgan\
        gfpgan\
        safety_checker\
```

### Running the Tests

Simply execute: `tox` (or `tox -q` for less noisy output)

This will take a while the first time as it installs all the dependencies.

If the tests run successfully images will be produced in the `images/` folder.

#### Running a specific test file

`tox -- -k <filename>` for example `tox -- -k test_initialisation`

### Directory Structure

`hordelib/pipeline_designs/`
Contains ComfyUI pipelines in a format that can be opened by the ComfyUI web app. These saved directly from the web app.

`hordelib/pipelines/`
Contains the above pipeline JSON files converted to the format required by the backend pipeline processor. These are converted from the web app, see _Converting ComfyUI pipelines_ below.

`hordelib/nodes/` These are the custom ComfyUI nodes we use for `hordelib` specific processing.

### Running ComfyUI Web Application

`tox -e comfyui`

Then open a browser at: http://127.0.0.1:8188/

### Designing ComfyUI Pipelines

Use the standard ComfyUI web app. Use the "title" attribute to name the nodes, these names become parameter names in the `hordelib`. For example, a KSampler with the "title" of "sampler2" would become a parameter `sampler2.seed`, `sampler2.cfg`, etc. Load the pipeline `hordelib/pipeline_designs/pipeline_stable_diffusion.json` in the ComfyUI web app for an example.

Save any new pipeline in `hordelib/pipeline_designs` using the naming convention "pipeline_\<name\>.json".

Convert the JSON for the model (see _Converting ComfyUI pipelines_ below) and save the resulting JSON in `hordelib/pipelines` using the same filename as the previous JSON file.

That is all. This can then be called from `hordelib` using the `run_image_pipeline()` method in `hordelib.comfy.Comfy()`

### Converting ComfyUI pipelines

In addition to the design file saved from the UI, we need to save the pipeline file in the backend format. This file is created in the `hordelib` project root named `comfy-prompt.json` automatically if you run a pipeline through the ComfyUI version embedded in `hordelib`. Running ComfyUI with `tox -e comfyui` automatically patches ComfyUI so this JSON file is saved.

### Build Configuration

The main config files for the project are: `pyproject.toml`, `tox.ini` and `requirements.txt`

### PyPi Publishing

_Will be automatic from the `releases` branch_

But currently:

1. checkout the `releases` branch
1. `tox` _make sure everything works_
1. `python build_helper.py` _builds the dist files_
1. `twine upload -r pypi dist/*`

### Standalone "clean" environment test from Pypi

Here's an example:

Start in a new empty directory. Create requirements.txt:
```
--extra-index-url https://download.pytorch.org/whl/cu117
hordelib
```

Create the directory `images/` and copy the `test_db0.jpg` into it.

Copy `run_controlnet.py` from the `hordelib/tests/` directory.

Build a venv:
```
python -m venv venv
.\venv\Scripts\activate
pip install -r requirements.txt
```

Run the test we copied:
```
python run_controlnet.py

The `images/` directory should have our test images.
```

### Updating the embedded version of ComfyUI

We use a ComfyUI version pinned to a specific commit, see `hordelib/consts.py:COMFYUI_VERSION`

To test if the latest version works and upgrade to it, from the project root simply:

1. `cd ComfyUI` _Change CWD to the embedded comfy_
1. `git checkout master` _Switch to master branch_
1. `git pull` _Get the latest comfyui code_
1. `git rev-parse HEAD` _Update the hash in `hordelib.consts:COMFYUI_VERSION`_
1. `cd ..` _Get back to the hordelib project root_
1. `tox` _See if everything still works_

Now ComfyUI is pinned to a new version.

### ComfyUI Patching

We need to patch the ComfyUI source code. It's only a small patch to:

1. Allow ComfyUI to find our custom nodes without copying files and folder around.
2. Allow make ComfyUI output some handy JSON we need for development purposes.

To create a patch file:
- Make the required changes to a clean install of ComfyUI and then run `git diff > yourfile.patch` then move the patch file to wherever you want to save it.

Note that the patch file _really_ needs to be in UTF-8 format and some common terminals, e.g. Powershell, won't do this by default. In Powershell to create a patch file use: `git diff | Set-Content -Encoding utf8 -Path yourfile.patch`

Patches can be applied with the `hordelib.install_comfyui.Installer` classes `apply_patch()` method.


CHANGELOG.md
## hordelib Changelog

## [v1.3.17](https://github.com/Haidra-Org/hordelib/compare/v1.3.16...v1.3.17)

2 June 2023

- fix: wrong check _check_for_refresh [`beddc3d`](https://github.com/Haidra-Org/hordelib/commit/beddc3d8042715e84f414ef1e55f62a61b5e9a6b)  (db0)

## [v1.3.16](https://github.com/Haidra-Org/hordelib/compare/v1.3.15...v1.3.16)

1 June 2023

- feat: faster way to retrieve curated list [`0c29034`](https://github.com/Haidra-Org/hordelib/commit/0c29034bbde91b0d4b37da6713f4f1cdb402887b)  (db0)
- lint: fix [`ba8c098`](https://github.com/Haidra-Org/hordelib/commit/ba8c098537cc8bd6f987061a94e1c9d0cd13c98c)  (db0)
- fix: remove some debug msg [`6699bc5`](https://github.com/Haidra-Org/hordelib/commit/6699bc5f75662e7f093e1aa083156af0faa8589d)  (db0)

## [v1.3.15](https://github.com/Haidra-Org/hordelib/compare/v1.3.14...v1.3.15)

1 June 2023

- feat: Added model rechecking/refreshing [`2bce317`](https://github.com/Haidra-Org/hordelib/commit/2bce317ca32aee18a63c9948758bd9dc77392658)  (db0)
- feat: allows SFW workers to ignore nsfw loras [`c30672e`](https://github.com/Haidra-Org/hordelib/commit/c30672ea5cd2f76d3b7ba3e5cfb917aa006f51c4)  (db0)
- feat: allow larger loras [`843b495`](https://github.com/Haidra-Org/hordelib/commit/843b495e02a8a4d3f20d72878ca1242064c3f984)  (db0)

## [v1.3.14](https://github.com/Haidra-Org/hordelib/compare/v1.3.13...v1.3.14)

31 May 2023

- feat: switch to curated list [`daa3be9`](https://github.com/Haidra-Org/hordelib/commit/daa3be964874846026ce2cd72ed903f95acf7c8c)  (db0)
- tests: reorganize lora tests [`1334852`](https://github.com/Haidra-Org/hordelib/commit/1334852eb1dc7af9aea137179aa9d3e195625c75)  (db0)
- fix: Ignore huge loras [`b83a201`](https://github.com/Haidra-Org/hordelib/commit/b83a20153911db82ec0c2dd6f0886ed9b5c8329e)  (db0)

## [v1.3.13](https://github.com/Haidra-Org/hordelib/compare/v1.3.12...v1.3.13)

31 May 2023

- working adhoc loras [`f63dd29`](https://github.com/Haidra-Org/hordelib/commit/f63dd29292eac55c3b1df80b042964c487aa24e9)  (db0)
- fix: Ignore non-existent loras [`4617bc4`](https://github.com/Haidra-Org/hordelib/commit/4617bc4e0a745f1a2c59152cdae4f8fda532dd60)  (db0)
- fix: avoid crashing when a lora is to be ignored [`8df9f96`](https://github.com/Haidra-Org/hordelib/commit/8df9f96e8956756047f4bac1a2069c3f7dd3ca4a)  (db0)

## [v1.3.12](https://github.com/Haidra-Org/hordelib/compare/v1.3.11...v1.3.12)

30 May 2023

- fix: loras don't always have triggers [`5298407`](https://github.com/Haidra-Org/hordelib/commit/52984073c713c0a03bfab7a1b6ff495429d4a793)  (db0)

## [v1.3.11](https://github.com/Haidra-Org/hordelib/compare/v1.3.10...v1.3.11)

29 May 2023

- fix: tighten up thread safety around the sampler [`#341`](https://github.com/Haidra-Org/hordelib/pull/341) (Jug)
- fix: don't return results of another job in certain corner cases. [`#339`](https://github.com/Haidra-Org/hordelib/pull/339) (Jug)

## [v1.3.10](https://github.com/Haidra-Org/hordelib/compare/v1.3.9...v1.3.10)

29 May 2023

- fix: handles lora name being sent as a string int [`8f871b5`](https://github.com/Haidra-Org/hordelib/commit/8f871b5ecdbddcc6b012709ba48247786b6137b2)  (db0)

## [v1.3.9](https://github.com/Haidra-Org/hordelib/compare/v1.3.8...v1.3.9)

29 May 2023

- feat: Add seeking loras by ID and unicode [`4d6bb2d`](https://github.com/Haidra-Org/hordelib/commit/4d6bb2d6f6e3c02982617041e3a73659c1fc9cb0)  (db0)
- fix: avoid crash when resetting adhoc loras [`213a743`](https://github.com/Haidra-Org/hordelib/commit/213a743cd9d1eae8da56b3e12c4f5bb3de64313c)  (db0)

## [v1.3.8](https://github.com/Haidra-Org/hordelib/compare/v1.3.7...v1.3.8)

27 May 2023

- fix: logging error with loading cnet [`#332`](https://github.com/Haidra-Org/hordelib/pull/332) (Jug)

## [v1.3.7](https://github.com/Haidra-Org/hordelib/compare/v1.3.6...v1.3.7)

27 May 2023

- feat: keeping some unused loras as adhoc [`c81f2da`](https://github.com/Haidra-Org/hordelib/commit/c81f2da7ecafb704b99b11e6a5edc96e896e39fc)  (db0)
- chore: re-added copyright notes to nataili ports [`568738d`](https://github.com/Haidra-Org/hordelib/commit/568738d68b63431fbd5db250b003166122f2defd)  (db0)
- feat: More robust tracking of lora downloads [`28ae6a1`](https://github.com/Haidra-Org/hordelib/commit/28ae6a1ecbeabc42ef0f4040008e4c7246a98f7c)  (db0)

## [v1.3.6](https://github.com/Haidra-Org/hordelib/compare/v1.3.5...v1.3.6)

26 May 2023

- tests: Added test for lora model_reference wipe [`eac465d`](https://github.com/Haidra-Org/hordelib/commit/eac465dd9d80f07e5878f749aa8b7a62d541229e)  (db0)
- fix: wipe reference only when valid [`8b11641`](https://github.com/Haidra-Org/hordelib/commit/8b11641de44e4ac71e4f9745c1e7735aa01e736d)  (db0)
- feat: add changelog link to release annoucement [`c577b8a`](https://github.com/Haidra-Org/hordelib/commit/c577b8a1268b01e334e19919412777b6299a9077)  (Jug)

## [v1.3.5](https://github.com/Haidra-Org/hordelib/compare/v1.3.4...v1.3.5)

25 May 2023

- fix: allow all types of downloads to display progress [`#324`](https://github.com/Haidra-Org/hordelib/pull/324) (Jug)
- fix: make index uses .png files [`#321`](https://github.com/Haidra-Org/hordelib/pull/321) (Divided by Zer0)

## [v1.3.4](https://github.com/Haidra-Org/hordelib/compare/v1.3.3...v1.3.4)

25 May 2023

- tests: Compare test images with expected output [`#319`](https://github.com/Haidra-Org/hordelib/pull/319) (Divided by Zer0)
- feat: add support for download progress indicators [`#318`](https://github.com/Haidra-Org/hordelib/pull/318) (Jug)
- ci: set IMAGE_DISTANCE_THRESHOLD [`7a09e1a`](https://github.com/Haidra-Org/hordelib/commit/7a09e1afa8746ee25fc54647f40bb40e55afcf4e)  (db0)
- doc: restore PR unit test image link [`bf00539`](https://github.com/Haidra-Org/hordelib/commit/bf005391d174be84de987b0fa96c697747a424e1)  (Jug)
- doc: remove link to PR image tests which were removed [`c21453e`](https://github.com/Haidra-Org/hordelib/commit/c21453e706a2d5d516046fadca473e0815c94bd4)  (Jug)


hordelib/benchmark.py
import os
import platform
import time
from io import BytesIO

import requests
from loguru import logger
from PIL import Image

from hordelib.utils.gpuinfo import GPUInfo

FACE_FIX_TEST_IMAGE = "https://github.com/jug-dev/hordelib/blob/main/images/test_facefix.png?raw=true"
CONTROLNET_TEST_IMAGE = "https://github.com/jug-dev/hordelib/blob/main/images/test_db0.jpg?raw=true"


def download_image(url):
    response = requests.get(url)
    if response.status_code == 200:
        image_data = BytesIO(response.content)
        pil_image = Image.open(image_data)
        return pil_image
    else:
        raise Exception(f"Failed to download image. Status code: {response.status_code}")


timings = {}


def delta(desc):
    global timings
    if desc not in timings:
        timings[desc] = {"start": time.time(), "end": None}
        return 0
    else:
        timings[desc]["end"] = time.time()
        return timings[desc]["end"] - timings[desc]["start"]


def get_os():
    os_name = platform.system()
    if os_name == "Linux":
        import distro

        distro_name = distro.name()
        distro_version = distro.version()
        os_name = f"{os_name} ({distro_name} {distro_version})"
    elif os_name == "Windows":
        win_version = "-".join(platform.win32_ver())
        os_name = f"{os_name} ({win_version})"
    elif os_name == "Darwin":
        mac_ver, _, _ = platform.mac_ver()
        os_name = f"{os_name} ({mac_ver})"
    else:
        os_name = "Unknown OS"
    return os_name


def main():
    if not os.getenv("AIWORKER_CACHE_HOME"):
        print("No model directory found. Environmental variable AIWORKER_CACHE_HOME is not set.")
        exit(1)

    try:
        gpu = GPUInfo().get_info()
        gpu_name = gpu["product"]
        gpu_vram = gpu["vram_total"]
    except Exception:
        gpu = "unknown"
        gpu_name = ""
        gpu_vram = ""

    delta("initialisation")
    import hordelib

    hordelib.initialise(setup_logging=True)
    delta("initialisation")

    from hordelib.horde import HordeLib
    from hordelib.shared_model_manager import SharedModelManager

    try:
        from hordelib._version import __version__
    except Exception:
        __version__ = "dev branch"

    generate = HordeLib()
    delta("model-manager-load")
    SharedModelManager.loadModelManagers(compvis=True, controlnet=True)
    delta("model-manager-load")
    SharedModelManager.manager.load("stable_diffusion")

    data = {
        "sampler_name": "k_euler",
        "cfg_scale": 7.5,
        "denoising_strength": 1.0,
        "seed": 123456789,
        "height": 512,
        "width": 512,
        "karras": True,
        "tiling": False,
        "hires_fix": False,
        "control_strength": 1.0,
        "clip_skip": 1,
        "control_type": None,
        "image_is_control": False,
        "return_control_map": False,
        "prompt": "an ancient llamia monster ### sea, lake, cartoon",
        "ddim_steps": 100,
        "n_iter": 1,
        "model": "stable_diffusion",
        "source_processing": "txt2img",
    }

    logger.warning("Benchmarking inference")
    # overhead
    delta("basic-inference-overhead")
    data["ddim_steps"] = 1
    pil_image = generate.basic_inference(data)
    if not pil_image:
        raise Exception("Image generation failed")
    last = delta("basic-inference-overhead")
    inference_overhead = last
    # inference
    attempts = [10, 20, 30, 40, 50, 100, 200, 300]
    max_iterations = 0
    for attempt in attempts:
        delta(f"basic-inference-{attempt}-steps")
        data["ddim_steps"] = attempt
        pil_image = generate.basic_inference(data)
        if not pil_image:
            raise Exception("Image generation failed")
        last = delta(f"basic-inference-{attempt}-steps")
        if last > 30:
            break
        max_iterations = attempt
    its = round(attempt / last, 1)
    its_raw = round(attempt / (last - inference_overhead), 1)

    logger.warning("Benchmarking model load/unload")
    delta("stable-diffusion-model-load-x3")
    SharedModelManager.manager.unload_model("stable_diffusion")
    SharedModelManager.manager.load("stable_diffusion")
    SharedModelManager.manager.unload_model("stable_diffusion")
    SharedModelManager.manager.load("stable_diffusion")
    SharedModelManager.manager.unload_model("stable_diffusion")
    SharedModelManager.manager.load("stable_diffusion")
    last = delta("stable-diffusion-model-load-x3")
    model_load = round(last / 3, 1)

    logger.warning("Benchmarking controlnet")
    # approximate pipeline overhead
    data["ddim_steps"] = 1
    data["source_image"] = download_image(CONTROLNET_TEST_IMAGE)
    data["control_type"] = "hed"
    data["source_processing"] = "img2img"
    delta("controlnet-overhead")
    pil_image = generate.basic_inference(data)
    cnet_overhead = round(delta("controlnet-overhead"), 2)
    # inference
    data["ddim_steps"] = max_iterations
    data["source_image"] = download_image(CONTROLNET_TEST_IMAGE)
    data["control_type"] = "hed"
    data["source_processing"] = "img2img"
    delta("controlnet-inference")
    pil_image = generate.basic_inference(data)
    last = delta("controlnet-inference")
    cnet_its = round(max_iterations / last, 1)
    cnet_raw_its = round(max_iterations / (last - cnet_overhead), 1)
    if not pil_image:
        raise Exception("Image generation failed")

    # Display Results
    print()
    print()
    print(f"--- hordelib {__version__} Benchmark ---")
    print(f"    GPU: {gpu_name} {gpu_vram}")
    print(f"     OS: {get_os()}")
    print()
    print("    Iterations per second @ 512x512:")
    print(f"{its:>{9}} basic inference (empirical)")
    print(f"{its_raw:>{9}} basic inference (theoretical)")
    print(f"{cnet_its:>{9}} controlnet inference (empirical)")
    print(f"{cnet_raw_its:>{9}} controlnet inference (theoretical)")
    print()
    print(f"{model_load:>{9}}s model load speed")
    print()
    print("    Details:")
    for name, data in timings.items():
        secs = round(data["end"] - data["start"], 2)
        print(f"{secs:>{9}}s {name}")
    print()


if __name__ == "__main__":
    main()
